### PR TITLE
Mark blobs as immutable on disk, closes #139

### LINKED ISSUE #139: chattr +i blobs, chattr -i blobs before deleting them
This is a defense-in-depth against layers being modified while on disk.

### PR DESCRIPTION
**Pull request by peusebiu**

### CODE REVIEW

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -809,6 +810,12 @@ func (is *ImageStore) FinishBlobUpload(repo string, uuid string, body io.Reader,
 		}
 	}
 
+	err = chattr.AddImmutable(src)
```
Comment (tych0) [inline pkg/storage/storage.go:813]: Looks like this library just execs `chattr`, if we're going to use a library it might be nice to use one that actually does the ioctls in case someone doesn't have e2fsprogs installed. Otherwise we can just exec the chattr ourselves and avoid the dep.

Comment (tych0) [reply pkg/storage/storage.go:813]: It also just fails silently if chattr isn't present :\

https://github.com/libopenstorage/openstorage/blob/master/pkg/chattr/chattr.go#L19

Comment (peusebiu) [reply pkg/storage/storage.go:813]: I see, this one seems to use ioctls: https://github.com/g0rbe/go-chattr/blob/master/chattr.go
Is this one ok?

Comment (tych0) [reply pkg/storage/storage.go:813]: Yeah, that looks good to me, thanks!

**Review by shimish2 — COMMENTED** [review]
Is it possible to write unit tests for this change ?

File: pkg/storage/storage.go [hunk 2]
```diff
@@ -774,7 +775,7 @@ func (is *ImageStore) FinishBlobUpload(repo string, uuid string, body io.Reader,
 	}
 
 	srcDigest, err := godigest.FromReader(f)
-	f.Close()
+	defer f.Close()
```
Comment (tych0) [inline pkg/storage/storage.go:778]: Heh, this seems like an unrelated bugfix?

Comment (tych0) [reply pkg/storage/storage.go:778]: Oh, no, I'm just misunderstanding, sorry.

File: pkg/storage/storage.go [hunk 3]
```diff
@@ -1042,6 +1055,19 @@ func (is *ImageStore) DeleteBlob(repo string, digest string) error {
...
+	f, err := os.Open(blobPath)
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", src).Msg("failed to open blob")
+		return err
+	}
+	// unset immutable attribute
+	err = chattr.UnsetAttr(f, chattr.FS_IMMUTABLE_FL)
+	if err != nil {
+		is.log.Error().Err(err).Str("blobPath", blobPath).Msg("Unable to remove immutable attribute from blob")
+		return err
```
Comment (tych0) [inline pkg/storage/storage.go:1067]: I think we leak f here, right? We need a defer f.Close() above

Comment (peusebiu) [reply pkg/storage/storage.go:1067]: Yes, that's right, thanks, I will update it.

**Review by tych0 — APPROVED** [review]

**Comment by peusebiu** [comment]
I'm getting ioctl: operation not permitted...
Do you have any idea how to fix it? 

```
{"level":"error","error":"ioctl: operation not permitted","dst":"/tmp/oci-repo-test626962402/dedupe1/blobs/sha256/4455e7d3c20e85d61685e7ce4fc1ea5e9b9054afea55ee152b68ae8e90e64947","caller":"/home/travis/gopath/src/github.com/anuvu/zot/pkg/storage/storage.go:821","message":"Unable to make blob immutable"}
```

**Comment by tych0** [comment]
You need to run the tests with the CAP_LINUX_IMMUTABLE capability, which I guess they're not by default.

**Comment by peusebiu** [comment]
I've looked into travis documentation for capabilities, but I didn't find anything.

**Comment by tych0** [comment]
Simplest thing might just be to wrap the tests in capsh or sudo; looks
like you could do that in /Makefile.bazel in the `build` target.

**Comment by shimish2** [comment]
Could you please remove this commit "Merge branch 'master' into immutable_blobs" and instead use rebase to sync with master.

**Comment by rchincha** [comment]
Failing with this

```
sudo --preserve-env=PATH env go test -tags extended -v -race -cover -coverpkg ./... -coverprofile=coverage.txt -covermode=atomic ./...

sudo: option '--preserve-env' doesn't allow an argument

usage: sudo -h | -K | -k | -V

usage: sudo -v [-AknS] [-g group] [-h host] [-p prompt] [-u user]

usage: sudo -l [-AknS] [-g group] [-h host] [-p prompt] [-U user] [-u user]

            [command]

usage: sudo [-AbEHknPS] [-r role] [-t type] [-C num] [-g group] [-h host] [-p

            prompt] [-u user] [VAR=value] [-i|-s] [<command>]

usage: sudo -e [-AknS] [-r role] [-t type] [-C num] [-g group] [-h host] [-p

            prompt] [-u user] file ...

Makefile:27: recipe for target 'test' failed

**Comment by peusebiu** [comment]
Thank you Ram, I was wondering what was the error. Where did you find it?

**Comment by shimish2** [comment]
> Thank you Ram, I was wondering what was the error. Where did you find it?

https://travis-ci.org/github/anuvu/zot/builds/745439850#L1809

**Comment by peusebiu** [comment]
Now I'm getting os.Link() permission denied when deduping...
```
{"level":"error","error":"link /tmp/oci-repo-test943257812/zot-2-test/blobs/sha256/3c72a8ed68140139e483fe7368ae4d9651422749e91483557cbd5ecf99a96110 /tmp/oci-repo-test943257812/zot-4-test/blobs/sha256/3c72a8ed68140139e483fe7368ae4d9651422749e91483557cbd5ecf99a96110: operation not permitted","blobPath":"/tmp/oci-repo-test943257812/zot-4-test/blobs/sha256/3c72a8ed68140139e483fe7368ae4d9651422749e91483557cbd5ecf99a96110","link":"/tmp/oci-repo-test943257812/zot-2-test/blobs/sha256/3c72a8ed68140139e483fe7368ae4d9651422749e91483557cbd5ecf99a96110","caller":"/home/travis/gopath/src/github.com/anuvu/zot/pkg/storage/storage.go:962","time":"2020-11-23T22:46:30.541225796Z","caller":"/home/travis/gopath/src/github.com/anuvu/zot/pkg/storage/storage.go:962","message":"dedupe: unable to hard link"}
```

**Comment by peusebiu** [comment]
I'm still getting operation not permitted when os.Link() in dedupeBlob function, even if I remove the immutable attr from blobs before linking.

It's hard for me to fix it because I'm still unable to run tests in any of my environs. If you have any suggestion let me know. I don't know what to change anymore.

**Comment by shimish2** [comment]
> I'm still getting operation not permitted when os.Link() in dedupeBlob function, even if I remove the immutable attr from blobs before linking.
> 
> It's hard for me to fix it because I'm still unable to run tests in any of my environs. If you have any suggestion let me know. I don't know what to change anymore.

Let me test on my local, please give me permission to pull your branch.

**Comment by peusebiu** [comment]
Hello @shimish2, I managed to run the tests in one vm, and this version of PR is passing on my vm, but is failing on travis. Can you check this one please? Thanks a lot.

**Comment by shimish2** [comment]
> Hello @shimish2, I managed to run the tests in one vm, and this version of PR is passing on my vm, but is failing on travis. Can you check this one please? Thanks a lot.

Sure, I also raised #152 but it also failing on travis.

**Comment by rchincha** [comment]
Now that we are planning to add multiple filesystem support, chattr may not work on all filesystems.

**Comment by tych0** [comment]
> Now that we are planning to add multiple filesystem support, chattr may not work on all filesystems.

Can't you just ignore EROFS and log ENOSYS (or EINVAL or whatever the error is)? Still seems like a useful protection, as most sane filesystems will support it...
